File: src/python/LeetCode150/buysellstocks_I/BestTimeToBuyAndSellStock1_SinglePass.py
def max_profit(prices):
    max_profit = 0
    min_price = float('inf')  # Initialize with a very large value

    for price in prices:
        if price < min_price:
            min_price = price  # Update the minimum price
        elif price - min_price > max_profit:
            max_profit = price - min_price  # Calculate and update the maximum profit

    return max_profit

# Example usage
if __name__ == "__main__":
    # Input: List of stock prices
    prices = [7, 1, 5, 3, 6, 4]
    print("Maximum Profit (Single Pass):", max_profit(prices))

File: src/java/LeetCode/buysellstocks_I/BestTimeToBuyAndSellStock1_SinglePass.java
import java.util.Arrays;

public class BestTimeToBuyAndSellStock1_SinglePass {
    public static int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE; // Initialize the minimum price to the largest possible value
        int maxProfit = 0; // Initialize the maximum profit to 0

        // Iterate through the array
        for (int price : prices) {
            // Update minPrice if the current price is lower than minPrice
            if (price < minPrice) {
                minPrice = price;
                //System.out.println(minPrice);
            }
            // Calculate profit if we sell at the current price and update maxProfit if it's higher
            else if (price - minPrice > maxProfit) {
                maxProfit = price - minPrice;
                //System.out.println("maxProfit : "+maxProfit);
            }
        }

        return maxProfit;
    }

    public static void main(String[] args) {
        int[] prices = {7, 1, 5, 3, 6, 4};
        System.out.println("Maximum Profit (Single Pass): " + maxProfit(prices));
    }
}